neps/nep-0384.md

```diff
@@ -0,0 +1,108 @@
+
+    connection = ConnectionEnd{TRYOPEN, counterpartyConnectionIdentifier, counterpartyPrefix,
+                               clientIdentifier, counterpartyClientIdentifier, version, delayPeriodTime, delayPeriodBlocks}
+    abortTransactionUnless(connection.verifyConnectionState(proofHeight, proofInit, counterpartyConnectionIdentifier, expectedConnectionEnd))
+    abortTransactionUnless(connection.verifyClientState(proofHeight, proofClient, clientState))
+    abortTransactionUnless(connection.verifyClientConsensusState(
+      proofHeight, proofConsensus, counterpartyClientIdentifier, consensusHeight, expectedConsensusState))
+
+    provableStore.set(connectionPath(identifier), connection)
+    addConnectionToClient(clientIdentifier, identifier)
+}
+
+// ConnOpenAck relays acceptance of a connection open attempt from chain B back to chain A (this code is executed on chain A).
+function connOpenAck(
+  identifier: Identifier,
+  clientState: ClientState,
+  version: string,
+  counterpartyIdentifier: Identifier,
+  proofTry: CommitmentProof,
+  proofClient: CommitmentProof,
+  proofConsensus: CommitmentProof,
+  proofHeight: Height,
+  consensusHeight: Height) {
+    abortTransactionUnless(consensusHeight < getCurrentHeight())
+    abortTransactionUnless(validateSelfClient(clientState))
+    connection = provableStore.get(connectionPath(identifier))
+    abortTransactionUnless((connection.state === INIT && connection.version.indexOf(version) !== -1)
+    expectedConsensusState = getHostConsensusState(consensusHeight) // we need to return a near header at this height.
+    expectedConnectionEnd = ConnectionEnd{TRYOPEN, identifier, getCommitmentPrefix(),
+                             connection.counterpartyClientIdentifier, connection.clientIdentifier,
+                             version, connection.delayPeriodTime, connection.delayPeriodBlocks}
+    abortTransactionUnless(connection.verifyConnectionState(proofHeight, proofTry, counterpartyIdentifier, expectedConnectionEnd)) // so we can verify that the counterparty chain has the correct header.
+    abortTransactionUnless(connection.verifyClientState(proofHeight, proofClient, clientState))
+    abortTransactionUnless(connection.verifyClientConsensusState(
+      proofHeight, proofConsensus, connection.counterpartyClientIdentifier, consensusHeight, expectedConsensusState))
+    connection.state = OPEN
+    connection.version = version
+    connection.counterpartyConnectionIdentifier = counterpartyIdentifier
+    provableStore.set(connectionPath(identifier), connection)
+}
+```
+
+The api `env::block_hash` allows us validate the full header submitted via contract calldata by comparing hashes.
+
+## Rationale and alternatives
+
+A few other designs were considered such as a function `env::block_header` which would return the full header at a given height, but was discouraged against a result of expensive db calls. This mapping of block numbers to headers are cheap enough to be stored on-chain.
```

Comment:
> This mapping of block numbers to headers are cheap enough to be stored on-chain.

What do you mean by that? I don’t think we want to store that on-chain in the sense that it’s stored in state of the latest block.

Reply:
> What do you mean by that?

Yes i did assume that they would be stored on-chain, would that not be the case? I worry about consensus failures if that is not the case.

Reply:
We might have different understanding of ‘on-chain’.  The data is on-chain in the sense that everyone agrees that if you’re in a branch whose head is block X than that uniquely determines hashes of all previous blocks.  However, it’s not stored ‘on-chain’ in the sense that this in formation isn’t immediately available in block X.  To get that information you have to fetch X’s parent, that block’s parent, that block’s parent etc.